server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2648,4 +2641,161 @@ public void afterRefresh(boolean didRefresh) throws IOException {
             refreshMetric.inc(System.nanoTime() - currentRefreshStartTime);
         }
     }
+
+    public Engine.Result applyResyncOperation(Translog.Operation operation) throws IOException {
+        return applyTranslogOperation(engineHolder.getEngineForResync(), operation, Engine.Operation.Origin.PEER_RECOVERY);
+    }
+
+    /**
+     * This method is called after each bulk of primary-replica resync operations are applied to this shard
+     * via {@link #applyResyncOperation(Translog.Operation)}.
+     *
+     * @param maxSeqNoFromPrimary the maximum sequence number from the newly promoted primary
+     */
+    public void afterApplyResyncOperationsBulk(long maxSeqNoFromPrimary) throws IOException {
+        if (maxSeqNoFromPrimary != UNASSIGNED_SEQ_NO) {
+            engineHolder.getEngineForResync().trimOperationsFromTranslog(operationPrimaryTerm, maxSeqNoFromPrimary);
+            engineHolder.adjustMinRequiredCheckpoint(maxSeqNoFromPrimary);
+        }
+        engineHolder.maybeActivatePendingEngine();
+    }
+
+    final boolean isEngineResetting() {
+        return engineHolder.hasPendingEngine();
+    }
+
+    final boolean canResetEngine() {
+        // TODO: do not reset the following shard
+        // other options to enable rollback on older indices:
+        // 1. send max_seq_no from the primary via replication requests
+        // 2. pass channel version when acquiring an index shard permit
+        return indexSettings.getIndexVersionCreated().onOrAfter(Version.V_6_4_0);
+    }
+
+    private Engine maybeResetEngineToSafeCommit() throws IOException {
+        synchronized (mutex) {
+            assert isEngineResetting() == false : "engine is being reset already";
+            final long lastSyncedGlobalCheckpoint = getLastSyncedGlobalCheckpoint();
+            final long currentMaxSeqNo = getEngine().getSeqNoStats(lastSyncedGlobalCheckpoint).getMaxSeqNo();
+            // Do not reset the current engine if it does not have any stale operations
+            // or the primary is on an old version which does not send the max_seq_no from the primary during resync.
+            // we need max_seq_no from the primary to activate the resetting engine when its history aligned with primary's.
+            if (currentMaxSeqNo == lastSyncedGlobalCheckpoint || canResetEngine() == false) {
+                return null;
+            } else {
+                engineHolder.makeActiveEngineSearchOnly();
+                final Engine pendingEngine = createNewEngine(newEngineConfig());
```

Comment:
this trims unsafe commits, possibly cleaning up segments that are referenced by the active search only engine?

Reply:
We open a directory reader in the constructor of a search-only engine and keeps that reader until we manually close the search-only engine. Holding that reader would prevent the segment files of the last commit from deleting during trimming unsafe commits.